### PR TITLE
PP-3362 Payment FAILED action email integration

### PR DESCRIPTION
**Pull request by DanailMinchev**
## WHAT

Added payment `FAILED` action email integration with `PAYMENT_FAILED` template

### CODE REVIEW

File: src/main/java/uk/gov/pay/directdebit/notifications/services/UserNotificationService.java [hunk 1]
```diff
@@ -68,6 +72,12 @@ public void sendPaymentConfirmedEmailFor(Transaction transaction, Payer payer, L
         return map;
     }
 
+    private HashMap<String, String> buildPaymentProblemPersonalisation() {
```
Comment (georgeracu) [inline src/main/java/uk/gov/pay/directdebit/notifications/services/UserNotificationService.java]: A better name for this method will make it much clearer. It looks like is putting DD Guarantee into a map, quite generic might be as simple as `buildPersonalisation()`

File: src/main/java/uk/gov/pay/directdebit/payments/services/TransactionService.java [hunk 2]
```diff
@@ -123,9 +123,12 @@ public PaymentRequestEvent paymentCreatedFor(Transaction transaction, Payer paye
         return paymentRequestEventService.registerPaymentCreatedEventFor(updatedTransaction);
     }
 
-    public PaymentRequestEvent paymentFailedFor(Transaction transaction) {
-        Transaction newTransaction = updateStateFor(transaction, SupportedEvent.PAYMENT_FAILED);
-        return paymentRequestEventService.registerPaymentFailedEventFor(newTransaction);
+    public PaymentRequestEvent paymentFailedFor(Transaction transaction, Payer payer, boolean sendEmail) {
+        if (sendEmail) {
```
Comment (georgeracu) [inline src/main/java/uk/gov/pay/directdebit/payments/services/TransactionService.java]: I believe this kind of conditionals should be avoided as you need to carry around that boolean for every call of this method. I think a cleaner way is to split lines `130-131` into a separate method `paymentFailedWithoutEmailFor(...)` and to have line `128` and a call to `paymentFailedWithoutEmailFor(...)` into another method `paymentFailedWithEmailFor(...)`. Then you will call each method based on your situation.

File: src/main/java/uk/gov/pay/directdebit/webhook/gocardless/services/handlers/GoCardlessMandateHandler.java [hunk 3]
```diff
@@ -60,14 +61,14 @@ public static GoCardlessMandateAction fromString(String type) {
                 GoCardlessMandateAction.CANCELLED, (Transaction transaction) -> {
                     Payer payer = payerService.getPayerFor(transaction);
                     if (!transactionService.findPaymentSubmittedEventFor(transaction).isPresent()) {
-                        transactionService.paymentFailedFor(transaction);
+                        transactionService.paymentFailedFor(transaction, payer, false);
```
Comment (georgeracu) [inline src/main/java/uk/gov/pay/directdebit/webhook/gocardless/services/handlers/GoCardlessMandateHandler.java]: Remove boolean and replace with appropriate method call

File: src/main/java/uk/gov/pay/directdebit/webhook/gocardless/services/handlers/GoCardlessMandateHandler.java [hunk 4]
```diff
@@ -60,14 +61,14 @@ public static GoCardlessMandateAction fromString(String type) {
...
                     }
                     return mandateService.mandateCancelledFor(transaction, payer);
                 },
                 GoCardlessMandateAction.FAILED, (Transaction transaction) -> {
                     Payer payer = payerService.getPayerFor(transaction);
                     mandateService.mandateFailedFor(transaction, payer);
-                    return transactionService.paymentFailedFor(transaction);
+                    return transactionService.paymentFailedFor(transaction, payer, false);
```
Comment (georgeracu) [inline src/main/java/uk/gov/pay/directdebit/webhook/gocardless/services/handlers/GoCardlessMandateHandler.java]: Remove boolean and replace with appropriate method call

File: src/main/java/uk/gov/pay/directdebit/webhook/gocardless/services/handlers/GoCardlessPaymentHandler.java [hunk 5]
```diff
@@ -63,7 +66,7 @@ public static GoCardlessPaymentAction fromString(String type) {
                 .put(GoCardlessPaymentAction.CONFIRMED, (Transaction transaction) ->
                         transactionService.findPaymentSubmittedEventFor(transaction)
                                 .orElseThrow(() -> new InvalidStateException("Could not find payment submitted event for payment request with id: " + transaction.getPaymentRequest().getExternalId())))
-                .put(GoCardlessPaymentAction.FAILED, transactionService::paymentFailedFor)
+                .put(GoCardlessPaymentAction.FAILED, (Transaction transaction) -> transactionService.paymentFailedFor(transaction, payerService.getPayerFor(transaction), true))
```
Comment (georgeracu) [inline src/main/java/uk/gov/pay/directdebit/webhook/gocardless/services/handlers/GoCardlessPaymentHandler.java]: Remove boolean and replace with appropriate method call

File: src/test/java/uk/gov/pay/directdebit/payments/services/TransactionServiceTest.java [hunk 6]
```diff
@@ -266,15 +267,32 @@ public void paymentPaidOutFor_shouldSetPaymentAsSucceeded_andRegisterAPaidOutEve
...
-    public void paymentFailedFor_shouldSetPaymentAsFailed_andRegisterAPaymentFailedEvent() {
+    public void paymentFailedFor_shouldSetPaymentAsFailed_andRegisterAPaymentFailedEvent_andDoNotSendEmail() {
 
         Transaction transaction = TransactionFixture
                 .aTransactionFixture()
                 .withState(PENDING_DIRECT_DEBIT_PAYMENT)
                 .toEntity();
 
-        service.paymentFailedFor(transaction);
+        service.paymentFailedFor(transaction, payer, false);
```
Comment (georgeracu) [inline src/test/java/uk/gov/pay/directdebit/payments/services/TransactionServiceTest.java]: Remove boolean and replace with appropriate method call.

File: src/test/java/uk/gov/pay/directdebit/payments/services/TransactionServiceTest.java [hunk 7]
```diff
@@ -266,15 +267,32 @@ public void paymentPaidOutFor_shouldSetPaymentAsSucceeded_andRegisterAPaidOutEve
...
+
+    @Test
+    public void paymentFailedFor_shouldSetPaymentAsFailed_andRegisterAPaymentFailedEvent_andSendEmail() {
+
+        Transaction transaction = TransactionFixture
+                .aTransactionFixture()
+                .withState(PENDING_DIRECT_DEBIT_PAYMENT)
+                .toEntity();
+
+        service.paymentFailedFor(transaction, payer, true);
```
Comment (georgeracu) [inline src/test/java/uk/gov/pay/directdebit/payments/services/TransactionServiceTest.java]: Remove boolean and replace with appropriate method call

File: src/test/java/uk/gov/pay/directdebit/webhook/gocardless/services/GoCardlessMandateHandlerTest.java [hunk 8]
```diff
@@ -174,17 +173,18 @@ public void handle_onAFailedMandateGoCardlessEvent_shouldRegisterAPayEventAsMand
         when(mockedGoCardlessService.findMandateForEvent(goCardlessEvent)).thenReturn(goCardlessMandateFixture.toEntity());
         when(mockedPayerService.getPayerFor(transaction)).thenReturn(payer);
         when(mockedMandateService.mandateFailedFor(transaction, payer)).thenReturn(paymentRequestEvent);
-        when(mockedTransactionService.paymentFailedFor(transaction)).thenReturn(paymentRequestEvent);
+        when(mockedTransactionService.paymentFailedFor(transaction, payer, false)).thenReturn(paymentRequestEvent);
```
Comment (georgeracu) [inline src/test/java/uk/gov/pay/directdebit/webhook/gocardless/services/GoCardlessMandateHandlerTest.java]: Remove boolean and replace with appropriate method call

File: src/test/java/uk/gov/pay/directdebit/webhook/gocardless/services/GoCardlessMandateHandlerTest.java [hunk 9]
```diff
@@ -174,17 +173,18 @@ public void handle_onAFailedMandateGoCardlessEvent_shouldRegisterAPayEventAsMand
...
 
         goCardlessMandateHandler.handle(goCardlessEvent);
 
-        verify(mockedTransactionService).paymentFailedFor(transaction);
+        verify(mockedTransactionService).paymentFailedFor(transaction, payer, false);
```
Comment (georgeracu) [inline src/test/java/uk/gov/pay/directdebit/webhook/gocardless/services/GoCardlessMandateHandlerTest.java]: Remove boolean and replace with appropriate method call

File: src/test/java/uk/gov/pay/directdebit/webhook/gocardless/services/GoCardlessMandateHandlerTest.java [hunk 10]
```diff
@@ -193,17 +193,18 @@ public void handle_onACancelledMandateGoCardlessEvent_shouldRegisterAPayEventAsM
         when(mockedPayerService.getPayerFor(transaction)).thenReturn(payer);
         when(mockedTransactionService.findPaymentSubmittedEventFor(transaction)).thenReturn(Optional.empty());
         when(mockedMandateService.mandateCancelledFor(transaction, payer)).thenReturn(paymentRequestEvent);
-        when(mockedTransactionService.paymentFailedFor(transaction)).thenReturn(paymentRequestEvent);
+        when(mockedTransactionService.paymentFailedFor(transaction, payer, false)).thenReturn(paymentRequestEvent);
```
Comment (georgeracu) [inline src/test/java/uk/gov/pay/directdebit/webhook/gocardless/services/GoCardlessMandateHandlerTest.java]: Remove boolean and replace with appropriate method call

File: src/test/java/uk/gov/pay/directdebit/webhook/gocardless/services/GoCardlessMandateHandlerTest.java [hunk 11]
```diff
@@ -193,17 +193,18 @@ public void handle_onACancelledMandateGoCardlessEvent_shouldRegisterAPayEventAsM
...
 
         goCardlessMandateHandler.handle(goCardlessEvent);
 
-        verify(mockedTransactionService).paymentFailedFor(transaction);
+        verify(mockedTransactionService).paymentFailedFor(transaction, payer, false);
```
Comment (georgeracu) [inline src/test/java/uk/gov/pay/directdebit/webhook/gocardless/services/GoCardlessMandateHandlerTest.java]: Remove boolean and replace with appropriate method call

File: src/test/java/uk/gov/pay/directdebit/webhook/gocardless/services/GoCardlessMandateHandlerTest.java [hunk 12]
```diff
@@ -215,14 +216,14 @@ public void handle_onACancelledMandateGoCardlessEvent_shouldRegisterAPayEventAsM
 
         goCardlessMandateHandler.handle(goCardlessEvent);
 
-        verify(mockedTransactionService, never()).paymentFailedFor(transaction);
+        verify(mockedTransactionService, never()).paymentFailedFor(transaction, payer, false);
```
Comment (georgeracu) [inline src/test/java/uk/gov/pay/directdebit/webhook/gocardless/services/GoCardlessMandateHandlerTest.java]: Remove boolean and replace with appropriate method call

File: src/test/java/uk/gov/pay/directdebit/webhook/gocardless/services/GoCardlessPaymentHandlerTest.java [hunk 13]
```diff
@@ -133,11 +140,11 @@ public void handle_onFailedPaymentGoCardlessEvent_shouldSetAPayEventAsFailed() {
         GoCardlessEvent goCardlessEvent = spy(GoCardlessEventFixture.aGoCardlessEventFixture().withAction("failed").toEntity());
 
         when(mockedGoCardlessService.findPaymentForEvent(goCardlessEvent)).thenReturn(goCardlessPaymentFixture.toEntity());
-        when(mockedTransactionService.paymentFailedFor(transaction)).thenReturn(paymentRequestEvent);
+        when(mockedTransactionService.paymentFailedFor(transaction, payer, true)).thenReturn(paymentRequestEvent);
```
Comment (georgeracu) [inline src/test/java/uk/gov/pay/directdebit/webhook/gocardless/services/GoCardlessPaymentHandlerTest.java]: Remove boolean and replace with appropriate method call

File: src/test/java/uk/gov/pay/directdebit/webhook/gocardless/services/GoCardlessPaymentHandlerTest.java [hunk 14]
```diff
@@ -133,11 +140,11 @@ public void handle_onFailedPaymentGoCardlessEvent_shouldSetAPayEventAsFailed() {
...
 
         goCardlessPaymentHandler.handle(goCardlessEvent);
 
-        verify(mockedTransactionService).paymentFailedFor(transaction);
+        verify(mockedTransactionService).paymentFailedFor(transaction, payer, true);
```
Comment (georgeracu) [inline src/test/java/uk/gov/pay/directdebit/webhook/gocardless/services/GoCardlessPaymentHandlerTest.java]: Remove boolean and replace with appropriate method call

File: src/main/java/uk/gov/pay/directdebit/payments/services/TransactionService.java (hunk 2 above)
Comment (DanailMinchev) [reply src/main/java/uk/gov/pay/directdebit/payments/services/TransactionService.java]: Good point!

**Pull request merged**